### PR TITLE
Refactor bloomfilter sublibrary

### PR DESCRIPTION
**Pull request by phadej**
- Remove fusion, insert and insertList
- Move to src/
- Use tasty
- Incremental hashing
- Parametrise over hashing scheme
- Expose internal modules, rename constuctors
- Use stylish-haskell
- Haddocks

### CODE REVIEW

File: bloomfilter/src/Data/BloomFilter.hs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+-- TODO
+--import "Data.BloomFilter.Hash" (cheapHashes)
+--
+--filter = create (cheapHashes 3) 1024 $ \mf -> do
+--           insertMB mf \"foo\"
+--           insertMB mf \"bar\"
+-- @
+--
+-- Note that the result of the setup function is not used.
+create :: Int        -- ^ number of hash functions to use
```
Comment (phadej) [inline bloomfilter/src/Data/BloomFilter.hs:99]: This is non-exported function, thus I didn't cleanup the haddock example (which is very dated)

File: bloomfilter/src/Data/BloomFilter.hs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                              if V.unsafeIndex (bitArray ub) idx
+                              then go (i + 1)
+                              else False
+{-# SPECIALIZE elemHashes :: CheapHashes a -> Bloom a -> Bool #-}
+
+-- | Query an immutable Bloom filter for non-membership.  If the value
+-- /is/ present, return @False@.  If the value is not present, there
+-- is /still/ some possibility that @False@ will be returned.
+notElem :: (Hashes h, Hashable a) => a -> Bloom' h a -> Bool
+notElem elt ub = notElemHashes (makeHashes elt) ub
```
Comment (dcoutts) [inline bloomfilter/src/Data/BloomFilter.hs:165]: Isn't this just
```suggestion
notElem elt ub = not (elem elt ub)
```
and so we don't need `notElemHashes` at all?

File: bloomfilter/src/Data/BloomFilter.hs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+--
+-- For a higher-level interface that is easy to use, see the
+-- "Data.BloomFilter.Easy" module.
+
+-- $performance
+--
+-- The implementation has been carefully tuned for high performance
+-- and low space consumption.
+
+-- $differences
```
Comment (dcoutts) [inline bloomfilter/src/Data/BloomFilter.hs:249]: :+1:

File: bloomfilter/src/Data/BloomFilter/Easy.hs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+-- functions for a Bloom filter, based on its expected maximum
+-- capacity and a desired false positive rate.
+--
+-- The false positive rate is the rate at which queries against the
+-- filter should return 'True' when an element is not actually
+-- present.  It should be a fraction between 0 and 1, so a 1% false
+-- positive rate is represented by 0.01.
+--
+-- This function will suggest to use a bloom filter of prime size.
+-- These theoretically behave the best.
```
Comment (dcoutts) [inline bloomfilter/src/Data/BloomFilter/Easy.hs:79]: Ok, that's new too.

**Review by dcoutts — APPROVED** [review]
This looks great.

File: bloomfilter/src/Data/BloomFilter.hs (hunk 2 above)
Comment (phadej) [reply bloomfilter/src/Data/BloomFilter.hs:165]: it's whether we trust GHC to do somewhat involved `case of case` transformation here.

**Review by dcoutts — APPROVED** [review]

**Pull request merged**